src/lib/when.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, reparentNodes, DirectiveFn, NodePart} from '../core.js';
+
+type PartCache = [NodePart, NodePart, DocumentFragment];
+const partCaches = new WeakMap<NodePart, PartCache>();
+
+/**
+ * Efficiently switches between two templates based on the given condition. The rendered
+ * content is cached, and re-used when switching conditions. Templates are evaluated
+ * lazily, so the passed values must be functions.
+ *
+ * While this directive can render any regular part, it makes the most sense when used
+ * with TemplateResulte since most other values are dirty checked already.
+ *
+ * Example:
+ *
+ * let checked = false;
+ *
+ * html`
+ *   when(checked, () => html`Checkmark is checked`, () => html`Checkmark is not checked`);
+ * `
+ *
+ * @param condition the condition to test against
+ * @param ifValue the content to render if
+ * @param elseValue
+ */
+export const when = (condition: boolean, ifValue: () => any, elseValue: () => any): DirectiveFn<NodePart> =>
+    directive((part: NodePart) => {
+      // Create cache if this was the first render
+      if (!partCaches.has(part)) {
```

Comment:
Using `Map#has()` causes there to be two map lookups in the hit case and three in the miss case. Use `Map#get()` instead and check for undefined:

```ts
let partCache = partCaches.get(part);
if (partCache === undefined) {
  partCache = {
    truePart: new NodePart(parentPart.templateFactory),
    falsePart: new NodePart(parentPart.templateFactory),
    container: document.createDocumentFragment(),
  };
  partCaches.set(part, partCache);
}
```

This is one lookup for a hit, two for a miss.